### ISSUE TITLE
VMAgent fails to match regex and drops target that should be kept

### ISSUE THREAD

**Issue by lllamnyp** [VictoriaMetrics/VictoriaMetrics#5297]
### Describe the bug

VMAgent uses a VMServiceScrape with a relabeling step
```yaml
      - source_labels:
          - __meta_kubernetes_service_label_prometheus_io_custom_target
          - __meta_kubernetes_service_label_prometheus_custom_target
        regex: ^(.+);|;(.+)$
        action: keep
```
to scrape a service with
```yaml
labels:
  prometheus.io/custom-target: smth
```
and drops the target. This target should be kept, as the concatenation of `source_labels` with a `;` separator returns `smth;`. Online regex parsers say that this is a match. Prometheus server **does not** drop these target for an equivalent ServiceMonitor.

Here's a more complete description of the scrape job with all prior relabeling steps included.

```yaml
  - job_name: serviceScrape/prometheus-operator/custom-service/0
    honor_labels: false
    kubernetes_sd_configs:
      - role: endpoints
    relabel_configs:
      - source_labels:
          - __meta_kubernetes_endpoint_address_target_kind
          - __meta_kubernetes_endpoint_address_target_name
        separator: ;
        regex: Node;(.*)
        replacement: ${1}
        target_label: node
      - source_labels:
          - __meta_kubernetes_endpoint_address_target_kind
          - __meta_kubernetes_endpoint_address_target_name
        separator: ;
        regex: Pod;(.*)
        replacement: ${1}
        target_label: pod
      - source_labels:
          - __meta_kubernetes_pod_name
        target_label: pod
      - source_labels:
          - __meta_kubernetes_pod_container_name
        target_label: container
      - source_labels:
          - __meta_kubernetes_namespace
        target_label: namespace
      - source_labels:
          - __meta_kubernetes_service_name
        target_label: service
      - source_labels:
          - __meta_kubernetes_service_name
        target_label: job
        replacement: ${1}
      - source_labels:
          - __meta_kubernetes_service_label_prometheus_io_custom_target
          - __meta_kubernetes_service_label_prometheus_custom_target
        regex: ^(.+);|;(.+)$
        action: keep
```

As of 2023-07-11 the same bug is observed in the [VictoriaMetrics playground](https://play.victoriametrics.com/select/accounting/1/6a716b0f-38bc-4856-90ce-448fd713e3fe/prometheus/graph/?#/relabeling?config=-+target_label%3A+%22test1%22%0A++source_labels%3A+%0A++++-+%22__meta_kubernetes_service_label_prometheus_io_custom_target%22%0A++++-+%22__meta_kubernetes_service_label_prometheus_custom_target%22%0A++regex%3A+%22.*%22%0A++replacement%3A+%22%240%22%0A-+target_label%3A+%22test2%22%0A++source_labels%3A+%0A++++-+%22__meta_kubernetes_service_label_prometheus_io_custom_target%22%0A++++-+%22__meta_kubernetes_service_label_prometheus_custom_target%22%0A++regex%3A+%22%28%28.%2B%29%3B%7C%3B%28.%2B%29%29%22%0A++replacement%3A+%22%240%22%0A&labels=%7B__meta_kubernetes_service_label_prometheus_io_custom_target%3D%22smth%22%7D).

### To Reproduce

Grab the config from the description and target a service with a single, non-empty label `prometheus.io/custom-target: smth`. It will be dropped. To look deeper, connect to the VMAgent http server and try it out in the target-relabel-debug tool. Equivalently, follow the link to the VictoriaMetrics playground above and check out the behavior there. If it were to work correctly, the relabeling steps would lead to the appearance of a label `test2`. This is not the case.

### Version

victoriametrics/vmagent:v1.94.0 docker image

**Comment by lllamnyp** [comment]
I've went through this with a debugger and narrowed it down to the regex helper utils in VM:

```
*github.com/VictoriaMetrics/VictoriaMetrics/lib/regexutil.PromRegex {
	prefix: "",
	suffix: ".+;|;.+",
	substrDotStar: "",
	substrDotPlus: ";",
```
The regex in the relabel configs gets boiled down to the form of `.+;.+` as indicated by the field `substrDotPlus`. This is a subtly different regex, than the one in the initial relabel config and does not match the input.

**Comment by lllamnyp** [comment]
It appears, that a workaround is the equivalent regex `(..*);|;(.*.)`. This way the helper utils fail to recognize that the expression is wrapped by `.*` or `.+` tokens.

**Comment by valyala** [comment]
@lllamnyp , thanks for filing the detailed bugreport and performing the initial investigation of the root cause for the bug!
Could you build `vmagent` from the commit 3e93fa61adede4396400d763ae5c17be43232be6 according to [these instructions](https://docs.victoriametrics.com/vmagent.html#how-to-build-from-sources) and verify whether it fixes the issue with improper handling of `.+;|;.+` regex inside [relabeling rules](https://docs.victoriametrics.com/vmagent.html#relabeling)?

**Comment by lllamnyp** [comment]
Glad to help! I'll give it a shot tomorrow.

**Comment by lllamnyp** [comment]
@valyala vmagent built from that commit no longer drops metrics on the offending relabeling step.

**Comment by valyala** [comment]
FYI, the bugfix has been included in [v1.87.11 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.11). It will be included in the new release as well.

**Comment by valyala** [comment]
FYI, this issue has been also fixed in [v1.93.8 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.8).

**Comment by valyala** [comment]
The bugfix for this issue has been included in [v1.95.0](https://docs.victoriametrics.com/CHANGELOG.html). Closing the issue as fixed then.
